Add typechecking indirection for Config

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #143152
* __->__ #143229


When we create a Config[T], we actually dynamically unbox this in the module, so lets have type checker believe that Config[T] creates a T. This enables proper typechecking support.

oulgen committed:
Add typechecking indirection for Config

[ghstack-poisoned]

c00w commented:
This is amazing, is there anyway to add some type checking tests?

oulgen commented:
@c00w well technically the next PR uses these which is some form of a typechecking test, but i dont think we traditionally had tests for this sort of thing

aorenste commented:
Positive tests will be added by uses of these.

I don't think we have any negative type-checking tests in the codebase. It's possible to do but I think we'd have to extend our test infra to support it.

aorenste commented on torch/utils/_config_module.py:
```diff
@@ -96,6 +97,29 @@ def __init__(
             ), f"justknobs only support booleans, {self.default} is not a boolean"
 
 
+if TYPE_CHECKING:
+
+    def Config(
+        default: T,
+        justknob: Optional[str] = None,
+        env_name_default: Optional[str] = None,
+        env_name_force: Optional[str] = None,
+        value_type: Optional[type] = None,
+    ) -> T:
+        ...
+
+else:
+
+    def Config(
+        default: T,
+        justknob: Optional[str] = None,
+        env_name_default: Optional[str] = None,
+        env_name_force: Optional[str] = None,
+        value_type: Optional[type] = None,
+    ) -> _Config[T]:
+        return _Config(default, justknob, env_name_default, env_name_force, value_type)
+
+
```
It would make the machinery a little more complex but these could be collapsed if we want to avoid arg duplication:
```
if TYPE_CHECKING:
  _ConfigT: TypeAlias[T] = T
else:
  _ConfigT: TypeAlias[T] = _Config[T]

def Config(...args...) -> _ConfigT[T]:
    if TYPE_CHECKING:
        ...
    else:
        return _Config(...args....)
```

aorenste commented on torch/utils/_config_module.py:
```diff
@@ -96,6 +97,29 @@ def __init__(
 
 
+if TYPE_CHECKING:
+
```
This could use a comment explaining why we're doing this crazy thing here...

oulgen committed:
None on "Add typechecking indirection for Config"



When we create a Config[T], we actually dynamically unbox this in the module, so lets have type checker believe that Config[T] creates a T. This enables proper typechecking support.



[ghstack-poisoned]
